mllib/src/main/scala/org/apache/spark/ml/lsh/LSH.scala

```diff
@@ -0,0 +1,270 @@
+
+  /**
+   * :: DeveloperApi ::
+   *
+   * Calculate the distance between two different keys using the distance metric corresponding
+   * to the hashFunction
+   * @param x One of the point in the metric space
+   * @param y Another the point in the metric space
+   * @return The distance between x and y in double
+   */
+  protected[ml] def keyDistance(x: KeyType, y: KeyType): Double
+
+  /**
+   * :: DeveloperApi ::
+   *
+   * Calculate the distance between two different hash Vectors. By default, the distance is the
+   * minimum distance of two hash values in any dimension.
+   *
+   * @param x One of the hash vector
+   * @param y Another hash vector
+   * @return The distance between hash vectors x and y in double
+   */
+  protected[ml] def hashDistance(x: Vector, y: Vector): Double = {
+    (x.asBreeze - y.asBreeze).toArray.map(math.abs).min
+  }
+
+  /**
+   * Transforms the input dataset.
+   */
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    transformSchema(dataset.schema, logging = true)
+    val transformUDF = udf(hashFunction, new VectorUDT)
+    dataset.withColumn($(outputCol), transformUDF(dataset($(inputCol))))
+  }
+
+  /**
+   * :: DeveloperApi ::
+   *
+   * Check transform validity and derive the output schema from the input schema.
+   *
+   * Typical implementation should first conduct verification on schema change and parameter
+   * validity, including complex parameter interaction checks.
+   */
+  override def transformSchema(schema: StructType): StructType = {
+    transformLSHSchema(schema)
+  }
+
+  /**
+   * Given a large dataset and an item, approximately find at most k items which have the closest
+   * distance to the item.
+   * @param dataset the dataset to look for the key
+   * @param key The key to hash for the item
+   * @param k The maximum number of items closest to the key
+   * @param distCol The column to store the distance between pairs
+   * @return A dataset containing at most k items closest to the key. A distCol is added to show
+   *         the distance between each record and the key.
+   */
+  def approxNearestNeighbors(dataset: Dataset[_], key: KeyType, k: Int = 1,
+                             distCol: String = "distance"): Dataset[_] = {
+    if (k < 1) {
+      throw new Exception(s"Invalid number of nearest neighbors $k")
+    }
+    // Get Hash Value of the key v
+    val keyHash = hashFunction(key)
+    val modelDataset = transform(dataset)
+
+    // In the origin dataset, find the hash value u that is closest to v
+    val hashDistUDF = udf((x: Vector) => hashDistance(x, keyHash), DataTypes.DoubleType)
+    val nearestHashDataset = modelDataset.select(min(hashDistUDF(col($(outputCol)))))
+    val nearestHashValue = nearestHashDataset.collect()(0)(0).asInstanceOf[Double]
+
+    // Filter the dataset where the hash value equals to u
+    val modelSubset = modelDataset.filter(hashDistUDF(col($(outputCol))) === nearestHashValue)
```

Comment:
Actually this does not work because number of elements with the same "hashDistCol" can be much larger than k. In that case, we are random selecting k elements of the same "hashDistCol" value.

To resolve the issue you mentioned, I am changing nearestHashValue to hashThreshold, which is the maximum "hashDistCol" for the top k elements.